base_rest/restapi.py

```diff
@@ -0,0 +1,250 @@
+            return validator.document
+        raise UserError(_("BadRequest %s") % validator.errors)
+
+    def to_response(self, service, result):
+        validator = self.get_cerberus_validator(service)
+        if validator.validate(result):
+            return validator.document
+        raise SystemError(_("Invalid Response %s") % validator.errors)
+
+    def to_openapi_query_parameters(self, service):
+        json_schema = self.to_json_schema(service)
+        parameters = []
+        for prop, spec in list(json_schema["properties"].items()):
+            params = {
+                "name": prop,
+                "in": "query",
+                "required": prop in json_schema["required"],
+                "allowEmptyValue": spec.get("nullable", False),
+                "default": spec.get("default"),
+            }
+            if spec.get("schema"):
+                params["schema"] = spec.get("schema")
+            else:
+                params["schema"] = {"type": spec["type"]}
+            if spec.get("items"):
+                params["schema"]["items"] = spec.get("items")
+            if "enum" in spec:
+                params["schema"]["enum"] = spec["enum"]
+
+            parameters.append(params)
+
+            if spec["type"] == "array":
+                # To correctly handle array into the url query string,
+                # the name must ends with []
+                params["name"] = params["name"] + "[]"
+
+        return parameters
+
+    def to_openapi_requestbody(self, service):
+        return {
+            "content": {"application/json": {"schema": self.to_json_schema(service)}}
+        }
+
+    def to_openapi_responses(self, service):
+        json_schema = self.to_json_schema(service)
+        return {"200": {"content": {"application/json": {"schema": json_schema}}}}
+
+    def get_cerberus_validator(self, service):
+        schema = None
+        if isinstance(self._schema, dict):
+            # schema is a cerberus schema
+            schema = self._schema
+        elif isinstance(self._schema, str):
+            # schema is a method name to call on service to get the schema or
+            schema = getattr(service, self._schema)()
+        if isinstance(schema, Validator):
+            return schema
+        if isinstance(schema, dict):
+            return Validator(schema, purge_unknown=True)
+        raise Exception(_("Unable to get cerberus schema from %s") % self._schema)
+
+    def to_json_schema(self, service):
+        schema = self.get_cerberus_validator(service).schema
+        return cerberus_to_json(schema)
+
+
+class Datamodel(RestMethodParam):
+    def __init__(self, name, is_list=False):
+        self._name = name
+        self._is_list = is_list
+
+    def from_params(self, service, params):
+        ModelClass = service.env.datamodels[self._name]
+        try:
+            return ModelClass.load(
+                params, many=self._is_list, unknown=marshmallow.EXCLUDE
+            )
```

Comment:
a `.dump()` is required here. It is not working when input json is contains a list with a dict inside.

Reply:
`dump()`???? I don't follow you. If your input is a list, you must declare your param's method as `Datamodel('my.rest.model', is_list=1)`

Reply:
I know, but it was not received properly on the rest function.

Reply:
My error defining the function. Sorry